### DISCUSSION TITLE
How can I make a program permanent on a Nucleo-L432KC

### DISCUSSION THREAD

**Discussion by garryp4** [micropython/micropython#14460] [STM32 / Pyboard]
I am new to Micropython and trying to learn how to make a program permanent on the micro memory so it executes when powered on, The last 15 or so years I have built a number of circuits based on 8 bit Microchip PIC's and programmed in Basic. None of the circuits needed more horsepower than the 8 bit micro's. Now I need more horsepower so bought a Nucleo-l432kc to learn on. I have Micropython loaded using the STCube programmer and can do REPL all day. So far I have spent about 4 days trying to find how to make a file system on the micro so I can load main.py and execute it when it's powered on. I have tried rshel, mpfshell, storing on the micro with Thonny, os.mount, but have not gotten anywhere. Am I even approaching this the correct way?

Any help would be greatly appreciated.

**Comment by AndreoBotelho** [comment]
On the linker file from the 256k flash it has there's 26k reserved to flash user storage, you can send files to it using shell ( I like to use thonny to send, edit and execute) or usb storage and import it from boot.py or main.py to execute ir automatically. There's too 2 spi buses that you can add a spi flash if you need store more than 26k data.

**Reply by garryp4** [reply]
I have had no success with Thonny, rshell, or mpfshell. How exactly is it done?
-------------------------------------------------------------------------------
Here is mpfshell:
E:\MicroPython\Original File>dir
 Volume in drive E has no label.
 Volume Serial Number is 76FE-FFD3

 Directory of E:\MicroPython\Original File

05/07/2024  10:07 AM    <DIR>          .
05/07/2024  10:07 AM    <DIR>          ..
04/14/2024  10:27 AM    <DIR>          .idea
04/14/2024  10:29 AM               305 boot.py
05/07/2024  10:07 AM               290 MAIN.PY
04/15/2024  09:07 AM         1,025,250 PYBV11-20240222-v1.22.2.hex
               3 File(s)      1,025,845 bytes
               3 Dir(s)  478,395,441,152 bytes free

E:\MicroPython\Original File>mpfshell com9
Connected to pyboard

** Micropython File Shell v0.9.3, sw@kaltpost.de **
-- Running on Python 3.12 using PySerial 3.5 --

mpfs [/]> put boot.py

Failed to create file: boot.py

mpfs [/]>
-------------------------------------------------------------------------------
Here is rshell:
E:\MicroPython\Original File>rshell -p com9
Traceback (most recent call last):
  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "e:\python312\scripts\rshell.exe\__main__.py", line 4, in <module>
  File "E:\Python312\Lib\site-packages\rshell\command_line.py", line 1, in <module>
    import rshell.main
  File "E:\Python312\Lib\site-packages\rshell\main.py", line 80, in <module>
    import readline
ModuleNotFoundError: No module named 'readline'

E:\MicroPython\Original File>
-----------------------------------------------------------------------------------
Here is Thonny error:
PROBLEM IN THONNY'S BACK-END: Exception while handling 'write_file' (OSError: Could not open file /main.py for writing, output:
Traceback (most recent call last):

  File "<stdin>", line 4, in <module>

OSError: [Errno 19] ENODEV).
See Thonny's backend.log for more info.
You may need to press "Stop/Restart" or hard-reset your MicroPython device and try again.


Process ended with exit code 1.
-------------------------------------------------------------------------------------
This code from Thonny executes on the NUCLEO-L432KC fine, I just can't store it in the Micro memory so it executes when it powers up:

import machine
import time

# Define LED pin
LED_PIN = machine.Pin('PB3', machine.Pin.OUT)

# Main loop
while True:
    # Turn LED on
    LED_PIN.on()
    # Delay for 1 second
    time.sleep(.2)
    # Turn LED off
    LED_PIN.off()
    # Delay for 1 second
    time.sleep(.2)
-------------------------------------------------------------------------------------

The NUCLEO-L432KC does not have a memory card slot. I chose this NUCLEO because it has a smaller micro. The circuit I will build doesn't need many pins.

Thanks

**Comment by AndreoBotelho** [comment]
You can save files on "/flash" drive, it appears that you are trying to
save on "/" root
Em sáb., 11 de mai. de 2024 11:56, garryp4 ***@***.***>
escreveu:

**Comment by garryp4** [comment]
how can I do that? Thonny only has an option for 'This Computer and Mycropython Device. I tried this in mpfshell:

E:\MicroPython\Original File>mpfshell com9
Connected to pyboard

** Micropython File Shell v0.9.3, sw@kaltpost.de **
-- Running on Python 3.12 using PySerial 3.5 --

mpfs [/]> put main.py /flash

Failed to create file: /flash

mpfs [/]>

**Comment by andrewleech** [comment]
Three official tool for this sort of thing is mpremote
https://docs.micropython.org/en/latest/reference/mpremote.html

`mpremote cp main.py :`

If this doesn't work, the filesystem on the device might need reformatting.

https://docs.micropython.org/en/latest/reference/filesystem.html

```
# STM32
import os, vfs, pyb
vfs.umount('/flash')
vfs.VfsLfs2.mkfs(pyb.Flash(start=0))
vfs.mount(pyb.Flash(start=0), '/flash')
os.chdir('/flash')
```

**Comment by garryp4** [comment]
This is from the Command Prompt:
E:\MicroPython\Original File>mpremote cp main.py :
cp main.py :
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
OSError: [Errno 19] ENODEV


E:\MicroPython\Original File>
----------------------------------------------------------
So in Thonny I entered this and ran it:
# STM32
import os, vfs, pyb
vfs.umount('/flash')
vfs.VfsLfs2.mkfs(pyb.Flash(start=0))
vfs.mount(pyb.Flash(start=0), '/flash')
os.chdir('/flash')

and got this error:
>>> %Run -c $EDITOR_CONTENT

MPY: sync filesystems
MPY: soft reboot
MPY: can't mount flash
Traceback (most recent call last):
  File "<stdin>", line 2, in <module>
ImportError: no module named 'vfs'
>>> 

I really do appreciate your help. I am so used to writing my code in basic, compiling, and flashing a hex file.

**Comment by andrewleech** [comment]
Ok, I don't use thonny, not sure if it's broken things.
However `MPY: can't mount flash` confirms the internal filesystem has been erased / corrupted somehow.

The instructions I pasted are for the current / latest micropython, `vfs` missing is a new module that your micropython must not have yet.

Try this
```
import os, pyb
os.umount('/flash')
os.VfsLfs2.mkfs(pyb.Flash(start=0))
os.mount(pyb.Flash(start=0), '/flash')
os.chdir('/flash')
```

**Comment by garryp4** [comment]
I get an error. Pasted from the Command Prompt window:

E:\Python312\Scripts>mpfshell com9
Connected to pyboard

** Micropython File Shell v0.9.3, sw@kaltpost.de **
-- Running on Python 3.12 using PySerial 3.5 --

mpfs [/]> help

Documented commands (type help <topic>):
========================================
EOF  cd     exec  get   lcd  lpwd  md    mput  mrm   put   pwd   rm
cat  close  exit  help  lls  ls    mget  mpyc  open  putc  repl

mpfs [/]> repl
>
*** Exit REPL with Ctrl+Q ***

MicroPython v1.22.2 on 2024-02-22; NUCLEO-L432KC with STM32L432KC
Type "help()" for more information.
>>> import os, pyb
>>> os.umount('/flash')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
OSError: [Errno 22] EINVAL
>>>

**Comment by garryp4** [comment]
I did a REPL to verify the serial connection is good:

E:\Python312\Scripts>mpfshell com9
Connected to pyboard

** Micropython File Shell v0.9.3, sw@kaltpost.de **
-- Running on Python 3.12 using PySerial 3.5 --

mpfs [/]> repl
>

*** Exit REPL with Ctrl+Q ***MicroPython
 v1.22.2 on 2024-02-22; NUCLEO-L432KC with STM32L432KC
Type "help()" for more information.
>>> print("hello")
hello
>>>
mpfs [/]> ^Z

E:\Python312\Scripts>

**Comment by garryp4** [comment]
I looked up what options I have with 'import os'. I found I can do some commands but not all. I'm not sure what it means tho. Another paste from the Command window:

E:\Python312\Scripts>mpfshell com9
Connected to pyboard

** Micropython File Shell v0.9.3, sw@kaltpost.de **
-- Running on Python 3.12 using PySerial 3.5 --

mpfs [/]> repl
>
*** Exit REPL with Ctrl+Q ***
Micr
oPython v1.22.2 on 2024-02-22; NUCLEO-L432KC with STM32L432KC
Type "help()" for more information.
>>> os.getcwd()
'/'
>>> os.stat('/')
(16384, 0, 0, 0, 0, 0, 0, 0, 0, 0)
>>> os.getenv('HOME')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'module' object has no attribute 'getenv'
>>>

**Comment by garryp4** [comment]
Now more information. I also happen to have a NUCLEO-G0B1RE. After loading the Micropython hex from GitHub, I opened Thonny and found that boot.py and main.py were already files on the micro. There is no external memory on this NUCLEO so the files need to be on the micro. Is andrewleech correct with"missing is a new module that your micropython must not have yet."?

**Reply by scruss** [reply]
maybe try pasting the lines in one at a time, and hit return at each so we can see which one caused an error.

It would also be really helpful is you enclosed your pasted text in three backticks before and after, as described in the [Welcome (Please read before posting)](https://github.com/orgs/micropython/discussions/9111) post. Your output is very hard to follow

**Comment by andrewleech** [comment]
As mentioned please try to post command prompt stuff in backticks to format them in a readable way.

From the commands I previously suggested, remove the line `os.umount('/flash')` and try the rest again. This particular one is just for when you have an existing filesystem mounted / being used. 

That os.getenv command you tried doesn't exist on micropython devices, there's no environment as such there. Only desktop / unix port would have access to an environment.

**Comment by garryp4** [comment]
I did read the three backticks but was using the apostrophe. My mistake. Anyway, here is the attempt. After the 'REPL' I did a reset on the board just to make sure it was connected and gave response. Then a print to make sure Micropython was alive. Then, as suggested by andrewleech, I skipped the unmount command. You can see VfsLfs2 upset it.  

E:\Python312\Scripts>mpfshell com9
Connected to pyboard

** Micropython File Shell v0.9.3, sw@kaltpost.de **
-- Running on Python 3.12 using PySerial 3.5 --

mpfs [/]> repl
>
*** Exit REPL with Ctrl+Q ***
MicroPytho
n v1.22.2 on 2024-02-22; NUCLEO-L432KC with STM32L432KC
Type "help()" for more information.
```
>>> print('hello')
hello
>>> import os, pyb
>>> os.VfsLfs2.mkfs(pyb.Flash(start=0))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'module' object has no attribute 'VfsLfs2'
>>>
```

The reason I'm using the L432KC flavor is the circuit I intend to build once I learn enough does not need even close to 64 pins. It would not even use 32 pins. And I need to keep the real estate to a minimum. I really do appreciate the help.

**Reply by scruss** [reply]
I have a very different board (F4 Discovery), and it doesn't seem to have the vfs module or the os.VfsLfs2.* anything in the system. It does have os.VfsFat.*, though.

I notice that mpfshell uses the remote filesystem mount point as the prompt:
```
mpfshell ttyACM1
Connected to pyboard

** Micropython File Shell v0.9.3, sw@kaltpost.de ** 
-- Running on Python 3.10 using PySerial 3.5 --

mpfs [/flash]> pwd
/flash
mpfs [/flash]> repl
>
MicroPython v1.22.2 on 2024-02-22; F4DISC with STM32F407
Type "help()" for more information.
>>> 
*** Exit REPL with Ctrl+] ***

mpfs [/flash]> 

```
This makes me think that your filesystem is at /, not /flash

**Comment by andrewleech** [comment]
Ah ok, looking at that board config it's a bit old, has LFS1 enabled but not 2
https://github.com/micropython/micropython/blob/master/ports%2Fstm32%2Fboards%2FNUCLEO_L432KC%2Fmpconfigboard.mk

Can try again with VfsLfs1 instead or VfsFat is also probably enabled, though it's not as robust as LFS

**Comment by garryp4** [comment]
That did it! I flashed a program that blinks the LED, power cycled it, and the LED is flashing. 

Thank you both so much. Now to get on with learning micropython. How can I tell how much flash memory there is?

**Comment by AndreoBotelho** [comment]
Use os.statvfs("/flash") that will return drive stats, use block size and
free blocks stats to calc free space on device.

Em seg., 13 de mai. de 2024 12:20, garryp4 ***@***.***>
escreveu:
